Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+                                docSnapshot.UpdateTime.GetValueOrDefault()));
+                        batchSize += 1;
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize == pageSize);
+
+            // Purge entries whose SlidingExpiration has passed.
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderBy("LastRefresh")
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    CacheDoc doc = docSnapshot.ConvertTo<CacheDoc>();
+                    if (doc.SlidingExpirationSeconds.HasValue)
+                    {
+                        var slidingExpiration =
+                            doc.LastRefresh.GetValueOrDefault()
+                            + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.Value);
+                        if (slidingExpiration < now)
+                        {
+                            writeBatch.Delete(docSnapshot.Reference,
+                                Precondition.LastUpdated(
+                                    docSnapshot.UpdateTime.GetValueOrDefault()));
+                            batchSize += 1;
+                        }
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize > 0);
+            _logger.LogTrace("End garbage collection.");
+        }
+
+        internal static string GetProjectId()
+        {
+            // Use the service account credentials, if present.
+            GoogleCredential googleCredential = Google.Apis.Auth.OAuth2
+                .GoogleCredential.GetApplicationDefault();
+            if (googleCredential != null)
+            {
+                ICredential credential = googleCredential.UnderlyingCredential;
+                ServiceAccountCredential serviceAccountCredential =
+                    credential as ServiceAccountCredential;
+                if (serviceAccountCredential != null)
+                {
+                    return serviceAccountCredential.ProjectId;
+                }
+            }
+            try

```

Comment:
Use `Platform.Instance()` for this instead? That has a `ProjectId` property you can use (and which I've just found a bug in for Cloud Run!)

Reply:
In which version of the library is the bug fixed?

Reply:
The Cloud Run bug will be fixed in 2.10.0-beta02 on Monday. At 2.9.0 there isn't support for Cloud Run at all.

Reply:
2.10.0-beta02 is now available. It doesn't do the "get it from the service credential" part though.